spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+        var i = cursor;
+
+        if (i >= parent.length)
+        {
+            action THROW_NEW("java.util.NoSuchElementException", []);
+        }
+
+        // #problem
+        //I don't know what to do with ConcurrentModificationException(); ?
+
+        cursor = i + 1;
+        lastRet = i;
+        result = action LIST_GET(self.parent.storage, lastRet);
+    }
+
+
+    fun previous (): Object
+    {
+        self.parent._checkForComodification(expectedModCount);
+        var i = cursor - 1;
+
+        if (i < 0)
+        {
+            action THROW_NEW("java.util.NoSuchElementException", []);
+        }
+
+        // #problem
+        //I don't know what to do with ConcurrentModificationException(); ?
+
+        cursor = i;
+        lastRet = i;
+        result = action LIST_GET(self.parent.storage, lastRet);
+    }
+
+
+    fun remove (): void
+    {
+        if (lastRet < 0)
+        {
+            action THROW_NEW("java.lang.IllegalStateException", []);
+        }
+
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
+
+        self.parent._deleteElement(lastRet);
+        cursor = lastRet;
+        lastRet = -1;
+        expectedModCount = self.parent.modCount;
+    }
+
+    fun set (e: Object): void
+    {
+        if (lastRet < 0)
+        {
+            action THROW_NEW("java.lang.IllegalStateException", []);
+        }
+
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
+
+        self.parent._setElement(lastRet, e);
+    }
+
+
+    fun add (e: Object): void
+    {
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
+
+        var i = cursor;
+        self.parent._addElement(self.parent.length, e);
+        cursor = i + 1;
+        lastRet = -1;
+        expectedModCount = self.parent.modCount;
+    }
+
+
+    fun forEachRemaining (action: Consumer): void
+    {
+        action NOT_IMPLEMENTED();
+    }
+
+}
+
+
+
+@final
+@implements("java.util.Spliterator")
+@WrapperMeta(
+    src="java.util.ArrayList$ArrayListSpliterator",
+    dst="org.utbot.engine.overrides.collections.ArrayList$UtArrayListSpliterator",
+    forceMatchInterfaces=true,
+)
+automaton ArrayListSpliterator: int(
+    var index: int,
+    var fence: int,
+    var expectedModCount: int
+)
+{
+
+    initstate Initialized;
+
+    shift Allocated -> Initialized by [
+        ArrayListSpliterator(int, int, int)
+    ];
+
+    shift Initialized -> self by [
+       // read operations
+       estimateSize,
+       characteristics
+
+       // write operations
+       trySplit,
+       tryAdvance,
+       forEachRemaining
+    ]
+
+    //constructors
+
+
+    constructor ArrayListSpliterator (origin: int, fence: int, expectedModCount: int)
```

Comment:
It is not possible (except reflection) to call the constructors of this class outside of the `ArrayList`.

Reply:
@dog-m As I understood you - I must add annotation "@private" ?

Reply:
I don't think there should even be constructors in the first place inside of such internal classes/automata. We have no way of making calls to it from LibSL perspective (it is a public/`fun`-like function after all).

Reply:
Fixed